### PR TITLE
barcrest/mpu4.cpp: Add IC4 port B shadowing of serial, 50Hz signal

### PR DESCRIPTION
**Pull request by james-wallace-ghub**
Clones promoted to working
--------------------------
The Crystal Maze (v1.3, Datapak) (MPU4 Video)
The New Crystal Maze Featuring Ocean Zone (v2.2, Datapak) (MPU4 Video) 
The Crystal Maze Team Challenge (v0.9, Datapak) (MPU4 Video)

### CODE REVIEW

File: src/mame/barcrest/mpu4.cpp [hunk 1]
```diff
@@ -179,7 +179,7 @@ IRQ line connected to CPU
  0F00-0F03 |R/W| D D D D D D D D | PIA6821 IC8
            |   |                 |
            |   |                 | port A
-           |   |                 |
+           |   |       	          |
```
Comment (ajrhacker) [inline src/mame/barcrest/mpu4.cpp:182]: There seems to have been a minor accident here.

Comment (james-wallace-ghub) [reply src/mame/barcrest/mpu4.cpp:182]: I think my IDE has eaten some spaces, let me see if I can fix it locally and push it back.

**Comment by ajrhacker** [comment]
There's still a stray tab character in there.

**Comment by james-wallace-ghub** [comment]
> There's still a stray tab character in there.

I hate VSCode, I've gone to old fashioned Notepad++ and cleared it out

**Comment by rb6502** [comment]
If you have/build the MAME tools, run srcclean on your files before you open a PR.  Vas srccleans the entire tree every release, but it avoids surprises with formatting if you do it yourself.

**Comment by james-wallace-ghub** [comment]
> If you have/build the MAME tools, run srcclean on your files before you open a PR. Vas srccleans the entire tree every release, but it avoids surprises with formatting if you do it yourself.

Good call I'm thinking of adding it to pre-commit or something like that so I can't ever forget it, as I've been doing full builds as a matter of course.

File: src/mame/barcrest/mpu4.h [hunk 2]
```diff
@@ -531,6 +531,8 @@ class mpu4_state : public driver_device
 	int m_t1 = 0;
 	int m_t3l = 0;
 	int m_t3h = 0;
+	int m_serial_output = 0;
+
```
[the remark this answers is not in the document]
Comment (james-wallace-ghub) [reply src/mame/barcrest/mpu4.h:535]: I don't think *any* members are properly registered from that list,but I think that's a case for another PR, as there's a few things I've noticed that need attention in these PIA handlers. I'll tidy the rest up, and then look through these variables to see what needs registering, and indeed, what is actually needed.

File: src/mame/barcrest/mpu4.cpp [hunk 3]
```diff
@@ -728,15 +724,14 @@ uint8_t mpu4_state::pia_ic4_portb_r()
 		if (m_optic_pattern & (1<<m_active_reel)) m_ic4_input_b |=  0x08;
 	}
 
-	if ( m_overcurrent )
+	if (m_low_volt_detect)
 	{
-		m_ic4_input_b |= 0x02;
+		if ( m_signal_50hz ) m_ic4_input_b |= 0x04;
 	}
```
Comment (cuavas) [inline src/mame/barcrest/mpu4.cpp:730]: I wasn’t objecting to the controlled statement on a line of its own – I was objecting to “English” parenthesis spacing `if (m_low_volt_detect)` being followed immediately by “European” parenthesis spacing `if ( m_signal_50hz )` just two lines later.

**Comment by james-wallace-ghub** [comment]
I grasp what you mean. I can't fix this before your next release so I'll put this in draft, resync it and fix the save state registration too, do it properly, so to say.

**Comment by james-wallace-ghub** [comment]
While I pull this together, I'd like to ask a style question. Where values can only have an on/off state, should these always be boolean, as opposed to int?

I'm noticing a few of these where they are single bit values that seem to have been made unsigned ints for convenience, and if I'm mapping this for state saving I'd like to reduce the footprint if I can.

**Comment by james-wallace-ghub** [comment]
This should cover everything, as well as figure out why we needed that special handler for the AY sound in one game. Turns out, the code was pointing to the wrong IC chip for the chipselect, and somehow that worked for some games by sheer dumb luck.

If there are any issues I'll address them, but I think this is a better foundation going forward.

**Comment by cuavas** [comment]
> While I pull this together, I'd like to ask a style question. Where values can only have an on/off state, should these always be boolean, as opposed to int?

Well, first of all, you should *never* be using `int` for anything that’s registered for save states.  You should be using explicitly sized integer types (either the abbreviated `u8`/`u16`/`s8`/`s16`/etc. or the standard `uint8_t`/`uint16_t`/`int8_t`/`int16_t`/etc.).  Using `int` etc. can break save state portability.

Now `bool` isn’t an explicitly sized type according to the C++ standard, but we get away with using it as though it is, because most compilers have adopted the Itanium C++ ABI, which *does* state that `bool` is always a single byte.  (I should write code to deal with this, but it would be complicated, and it isn’t actually causing a problem.  Note that the C99 `_Bool` which is aliased as `bool` in `<stdbool.h>` is *not* always a single byte – it’s four bytes on PowerPC for example.)

I wouldn’t be too worried about save state footprint.  The size of your Boolean variables is likely to be dwarfed by the size of RAM anyway.  One byte of four bytes won’t make a significant difference.  If you’re really worried, `u8` or `s8` are never larger than a `bool` anyway.

As for whether you use `bool` or some integer type, it partly comes down to how you plan to use it.  Are you going to be doing mathematical or bitwise operations on it where values 1 and 0 would be convenient?  Use an integer type.  Are you only going to be using it in Boolean context?  A `bool` is harmless.

**Review by cuavas — CHANGES_REQUESTED** [review]
Please change the `int` member variables that are registered for save states to explicitly sized integer types to help maintain save state portability.

**Comment by cuavas** [comment]
> Ok, I understand this. Pretty much everything here is representing a single byte of data, so it can all be unsigned.

That should be `u8`, not `unsigned` – using `unsigned` is as bad as using `int` as it isn’t explicitly sized.

**Comment by james-wallace-ghub** [comment]
> Please change the `int` member variables that are registered for save states to explicitly sized integer types to help maintain save state portability.



> Please change the `int` member variables that are registered for save states to explicitly sized integer types to help maintain save state portability.

This is done now, as well as a few other small cleanups. Hopefully this should do it, I can only test Windows

**Comment by Tafoid** [comment]
I'm finally getting to give this a test run for issues and there are.... MANY snapdiffs denoting regression.  Not that there are a lot of hardcore users of the fruities..  perhaps @james-wallace-ghub can take a look at these and determine if expected to need to be fixed?  There are way too many (over 3k m4* snapdiffs).

Available [HERE](https://mameload.mameworld.info/mpu4snapdiffs.7z)

**Comment by james-wallace-ghub** [comment]
> I'm finally getting to give this a test run for issues and there are.... MANY snapdiffs denoting regression. Not that there are a lot of hardcore users of the fruities.. perhaps @james-wallace-ghub can take a look at these and determine if expected to need to be fixed? There are way too many (over 3k m4* snapdiffs).
> 
> Available [HERE](https://mameload.mameworld.info/mpu4snapdiffs.7z)

This is brilliant, thanks - I'll check this evening. I've been trying to look at the regressions over time, and something like this is ideal. Are these batch tools around, I'd love to do this with a few version of the other drivers?

**Pull request merged**